File: backtracking/1079_letter_tile_possibilities/1079_letter_tile_possibilities.go
package main

func numTilePossibilities(tiles string) int {
	used := make([]bool, len(tiles))
	m := make(map[string]struct{})
	backtrack("", used, tiles, m)
	return len(m)
}

func backtrack(now string, used []bool, str string, m map[string]struct{}) {
	if len(now) > 0 {
		m[now] = struct{}{}
	}
	if len(now) == len(str) {
		return
	}
	for i := 0; i < len(str); i++ {
		if used[i] {
			continue
		}
		used[i] = true
		backtrack(now+string(str[i]), used, str, m)
		used[i] = false
	}
}

func main() {
	tiles := "AAB"
	res := numTilePossibilities(tiles)
	print(res)
}

File: backtracking/1079_letter_tile_possibilities/1079_letter_tile_possibilities.py
class Solution:
    def numTilePossibilities(self, tiles: str) -> int:
        record = [0] * 26
        for tile in tiles:
            record[ord(tile)- ord('A')] += 1
        
        def dfs(record):
            s = 0
            for i in range(26):
                if not record[i]:
                    continue
                record[i] -= 1
                s += dfs(record) + 1
                record[i] += 1
            return s
        return dfs(record)


tiles = "AAB"
res = Solution().numTilePossibilities(tiles)
print(res)